designs/helm-chart-autogenerate-plugin.md

```diff
@@ -0,0 +1,213 @@
+| Authors       | Creation Date | Status      | Extra |
+|---------------|---------------|-------------|---|
+| @name | date | Implementeble | - |
+
+# Auto-generate Helm Chart Plugin
+===================
+
+The proposal is to add a new alpha command to auto-generate a helm chart for a given kustomize directory created by kubebuilder.
```

Comment:
```suggestion
This proposal aims to introduce an optional mechanism that allows users to generate a Helm Chart from their Kubebuilder-scaffolded project. This will enable them to effectively package and distribute their solutions.
```

Here is more about the goal
How to achieve that will ihmo we need to add in the technical details section.

Reply:
let me try to understand, the  **Command** type is to add a new subcommand such as `kubebuilder generate-helm-chart`. And the **Plugin** type is to add a new internal plugin such as `kubebuilder init --plugins=helmchart/v1alpha`.
Is that what you mean? 

If so, I agree with the **Plugin** type. Users can get a default layout of helm chart from the [existing kustomize configurations](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4/config/default/kustomization.yaml) and [project configurations](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4/PROJECT).

As for synchronization with the kustomize directory, I think the main cost is **Adding/Deleting the CRDs**. However, we could leverage the kustomize and project configurations to generate a new helm chart once runs the command `kubebuilder init --plugins=helmchart/v1alpha`. Also, If there is a generated helm chart, the plugin will recognize it and   not affect the customizing part.

To make the above feature more clear, we can split it into two command `kubebuilder init` and `kubebuilder edit`.

Reply:
I have a concern, which I believe touches this subject. 
 
When we add a new marker for a CRD, and run `make manifests`, it generates new changes to the CRD in `config/crds`. When we add RBAC markers, it changes the Roles in the `config/rbac`. I'm talking about CRDs and RBAC, but could really be anything that changes the kustomize bases, and consequently should also changes the chart.

So I think that ideally, when we run `make manifests`, these new changes should also be reflected in `deploy/helm-chart/crds` (for CRD changes) and `deploy/helm-chart/templates` (for RBAC). @dashanji @camilamacedo86 do you agree this is a goal of ours?

And if so, does either the plugin or command affect this flow? Would we still be able to change the Makefile to include this new behavior regardless of it being a plugin or a command?

Reply:
> So I think that ideally, when we run make manifests, these new changes should also be reflected in deploy/helm-chart/crds (for CRD changes) and deploy/helm-chart/templates (for RBAC). @dashanji @camilamacedo86 do you agree this is a goal of ours?

Agree.

> And if so, does either the plugin or command affect this flow? Would we still be able to change the Makefile to include this new behavior regardless of it being a plugin or a command?

I think users are supposed to control the flow. First, generate a new kustomize directory by running `make manifests`, then use the plugin or command to generate a helm chart from the new kustomize directory. In my opinionYour concern is about how helm charts are consistent with kustomize manifests. Can we use the idea of ​​patch or merge to solve

Reply:
> I think users are supposed to control the flow. First, generate a new kustomize directory by running make manifests, then use the plugin or command to generate a helm chart from the new kustomize directory.

Ah ok, I was initially thinking of just running `make manifests`, not by running the plugin/command again. And that once we run the command/plugin once, it changes the Makefile to make sure `make manifests` also controls the helm chart.

I think this would be easier for users so that they don't forget to run an additional command to sync the chart, you know? I know I surely would forget many times :grin: 

How does that sound?

> In my opinion, your concern is about how helm charts are consistent with the customize manifests. Can we use the idea of ​​PATCH or MERGE (like kubectl patch) to solve such concern？

That wasn't my main concern here, but I agree that PATCH and MERGE will be good a good way to try to prevent undoing user customizations.

Reply:
> Ah ok, I was initially thinking of just running make manifests

That is exactly what I am proposing. :-)  Here's the workflow:

- Execute kubebuilder init | edit.
- During this step, inform the plugin that you're looking to use the helm chart.
- This action will scaffold a specific target within the Makefile tailored for this scenario.
- Hence, when you run make manifests or generate commands, the CRDs will automatically sync with the helm chart.